### PR TITLE
Added protected roles behavior

### PR DESCRIPTION
**Pull request by yanniboi**
by yanniboi: Added protected roles behavior for excluding roles during acquisition.

### CODE REVIEW

File: src/AcquisitionService.php [hunk 1]
```diff
@@ -171,6 +179,13 @@ protected function findMatch(array $values, array &$context) {
       $query->exists('name');
     }
 
+    // If we have the don't have the include protected roles behaviour, ensure
+    // protected roles are excluded.
+    if ($context['behavior'] ^ self::BEHAVIOR_INCLUDE_PROTECTED_ROLES) {
+      $roles = $this->settings->get('acquisitions.protected_roles');
+      $query->condition('roles', $roles, 'NOT IN');
```
Comment (andrewbelcher) [inline src/AcquisitionService.php]: We will need to check that we don't have no roles (empty array), as that will break the query.

File: src/DecoupledAuthSettingsForm.php [hunk 2]
```diff
@@ -73,6 +73,16 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#states' => ['visible' => ['input[name="acquisitions[registration]"' => ['value' => '1']]],
     ];
 
+    $roles = array_map(array('\Drupal\Component\Utility\Html', 'escape'), user_role_names(TRUE));
```
Comment (andrewbelcher) [inline src/DecoupledAuthSettingsForm.php:76]: Is there an injectable way to do this that isn't a huge amount of work?

**Review by andrewbelcher — CHANGES_REQUESTED** [review]
We need to update `config/schema/decoupled_auth.schema.yml` as well with the new `protected_roles` schema.

It would also be good to add some test coverage? `AcquisitionApiTest::testProtectedRoles` which tests:
- config default, no behaviour
- config default, use ignore roles behaviour
- config empty, works properly

File: src/DecoupledAuthSettingsForm.php (hunk 2 above)
Comment (yanniboi) [reply src/DecoupledAuthSettingsForm.php:76]: Yup lets leave this for now as I copied it from AccountForm

**Comment by andrewbelcher** [comment]
Due to https://www.drupal.org/node/2694381 (or related at least)

```
    // If we have the don't have the include protected roles behaviour, ensure
    // protected roles are excluded.
    if ($context['behavior'] ^ self::BEHAVIOR_INCLUDE_PROTECTED_ROLES) {
      $roles = $this->settings->get('acquisitions.protected_roles');
      $query->condition('roles', $roles, 'NOT IN');
    }
```

will have to be

```
    // If we have the don't have the include protected roles behaviour, ensure
    // protected roles are excluded.
    if ($context['behavior'] ^ self::BEHAVIOR_INCLUDE_PROTECTED_ROLES) {
      $roles = $this->settings->get('acquisitions.protected_roles');
      $or = $query->orConditionGroup();
      $or->condition('roles', $roles, 'NOT IN');
      $or->notExists('roles');
      $query->condition($or);
    }
```

**Review by andrewbelcher — APPROVED** [review]

**Pull request merged**